### ISSUE TITLE
Can't build after clang submodule added

### ISSUE THREAD

**Issue by wilsonk** [Syniurge/Calypso#30]
Hello Elie,

I still can't build Calypso even after the last couple changes to help with the new clang submodule. I rebuilt llvm in debug mode (which is going to really suck for build times and the QT demo), added libc++ and libc++-abi, since that seemed to be required??? Not absolutely sure on that one. Now I get this error:

```
[  0%] Built target impcnvgen
[  0%] Built target idgen
[ 16%] Built target LDCShared
[ 16%] Performing configure step for 'clang'
-- Found LLVM_CONFIG as /usr/local/bin/llvm-config
CMake Error at CMakeLists.txt:62 (if):
  if given arguments:

    "STREQUAL" "Debug" "AND" "NOT" "Debug+Asserts" "STREQUAL" "Debug"

  Unknown arguments specified

-- Configuring incomplete, errors occurred!
```

Maybe my version of cmake isn't new enough? cmake-2.8.12.2? I checked into this a little (after a bunch of errors with llvm stuff) and couldn't see what was wrong. 

'Frustrated' would be the word for my state of mind at this point, because I would like to try out the new Ogre demo and the build process really needs to be easier than this. I think the install instructions on the website need updating ...and with absolutely dead easy cut-n-paste instructions or people are going to get frustrated and not try Calypso out.

Sorry to be a nag, but it takes a good 30 mins to rebuild llvm/libc++, etc. and I have tried at least 4-5 different combinations over the last couple days and still don't have a working Calypso binary :( 

Thanks,
Kelly

P.S. Btw, I can't use the Ubuntu binaries you have spoken of for llvm/clang, because I prefer not to be on the absolute bleeding edge with my linux installs, so I am on 14.04 and not 15.04. I believe this would be the norm, as running non-LTS versions of Ubuntu is a little rare. I literally can't upgrade because the version of linux on my Calypso machine (which is non production) doesn't have an upgrade path to non-LTS versions.

**Comment by Syniurge** [comment]
Hi Kelly, sorry this was all due to missing "", it should be fixed now.

libc++ isn't needed at all, and if you want LLVM in `Release` or `RelWithDebInfo` mode for faster compilation times then you just need to build Calypso in `Release` or `RelWithDebInfo` mode too.

Other than that building Calypso is no different from before.

**Comment by wilsonk** [comment]
Ok, things built finally after the missing "" was fixed. I was still getting an error because I didn't do a 'git submodule update --init', which seems like it is required now for the clang submodule and not just the ldc submodules? But then things worked fine.

I would say that the build instructions should use the 'git submodule update --init' line above for the llvm/clang section, as well as the ldc section, if it is required to build everything. I would even highlight it because forgetting to update submodules can lead to some pretty cryptic errors.

The Ogre3d example built here also. It looks great and runs quite well even in debug mode. Amazing that 1273 object modules are built to get the demo working!!!! 

Thanks,
Kelly

**Comment by wilsonk** [comment]
Hello Elie,

I assume you know but those last few commits for MinGW and Twine have caused several libstdc++ and the ogre/qt demos to not compile anymore.

```
Instruction does not dominate all uses!
  %coerce = alloca %"class.Ogre::ColourValue", align 4
  %70 = bitcast %"class.Ogre::ColourValue"* %coerce to { <2 x float>, <2 x float> }*
Instruction does not dominate all uses!
  %coerce = alloca %"class.Ogre::ColourValue", align 4
  %76 = bitcast %"class.Ogre::ColourValue"* %coerce to i8*, !dbg !5842
Instruction does not dominate all uses!
  %coerce = alloca %"class.Ogre::ColourValue", align 4
  %101 = bitcast %"class.Ogre::ColourValue"* %coerce to { <2 x float>, <2 x float> }*
Instruction does not dominate all uses!
  %coerce = alloca %"class.Ogre::ColourValue", align 4
  %107 = bitcast %"class.Ogre::ColourValue"* %coerce to i8*, !dbg !5848
LLVM ERROR: Broken function found, compilation aborted!
```

I think they are all similar errors to this one above. The gdal sample works, at least.

**Comment by Syniurge** [comment]
That's strange, I don't see what change could have triggered this and I can't reproduce your errors, libstdc++ samples and the qt5 demo still compile here.

Does this happen on Ubuntu? Did you cleanup the calypso_cache files as usual? If yes could you try to revert a824bf53898c8bd69f20555780fc3b6df3ebb810?

But really no idea what could be happening.

BTW not related but another thing I forgot to tell you, if there's been changes to the `clang` submodule and you update it e.g with:

```
    git pull --recurse-submodules
```

Then before rebuilding Calypso you need to delete `build/clang-prefix/src/clang-stamp/clang-build` to trigger a Clang rebuild.

It's because cmake's ExternalProject doesn't do any check, and actually it's adding more confusion so I'm going to change this now.

**Comment by wilsonk** [comment]
Not completely sure why things are working now with a Release build of llvm/clang and calypso, (because I went and rebuilt everything from scratch)...but it IS working again now.

I think at least part of the problem was that clang wasn't rebuilding. But there also might have been a problem with the way I installed llvm/clang in debug mode (I have a debug version and a release version of llvm/clang/libc++/etc. in separate directories so that I don't have to rebuild llvm again!!!). I might have only installed part of llvm debug over top of the previous release install. Not too sure, but I think I have it worked out now. 

I did try to build and run after just removing the calypso_\* files, but that wasn't a problem (just so you know). I had also tried to revert that commit a824bf5, but that wasn't a problem either, it seems.

Anyways, I'll keep you informed if something breaks again here, but hopefully I have things worked out now.

Thanks,
Kelly
